Make `ocsf::derive` also map enum lists

jachris opened the pull request:
## 🔍 Problem

`ocsf::derive` derives scalar OCSF enum sibling fields, but skipped enum fields wrapped in lists. This left nested list enums such as DNS answer `flag_ids`/`flags` only partially normalized.

## 🛠️ Solution

- Derive enum siblings for `list<int>` and `list<string>` fields in both directions.
- Preserve sliced list-array offsets when rebuilding derived list fields.
- Validate bidirectional list pairs element-wise when both enum and sibling lists are present.
- Add regression coverage for forward, reverse, bidirectional, multi-list, and sliced-list derivation.

## 💬 Review

Focus on the Arrow list offset handling in `series::list_values()` and `make_list_series()`, plus the mixed scalar/list dispatch in `ocsf::derive`.

jachris committed:
Make `ocsf::derive` also map enum lists

mavam committed:
Credit changelog co-author

Assisted-by: GPT-5.1 Codex (pi)

mavam committed:
Cover OCSF enum list edges

Add regression coverage for null outer DNS flag lists and mismatched flag_ids/flags lengths. These cases exercise the list-slot preservation and diagnostic paths added for enum-list derivation.

Assisted-by: ChatGPT (pi)

mavam committed:
Keep OCSF enum lists column-wise

Require bidirectional enum lists to have the same Arrow list layout before deriving over their flattened values. This keeps null and empty outer lists distinct instead of repairing them row-wise.

Assisted-by: ChatGPT (pi)

IyeOnline commented on libtenzir/src/series.cpp:
```diff
@@ -76,8 +87,20 @@ auto make_record_series(std::span<const series_field> fields,
+        std::move(offsets), values.array, std::move(null_bitmap),
+        origin.null_count(), 0),
+    };
+  }
```
I would like to not have this in here. Its trying to re-use this function for an entirely different purpose in a way that is not generally safe.

There really should be two functions:

* Current `make_list_series` gets renamed `dangerously_rejoin_list_series`, but stays otherwise unchanged, including its warning comments.
* The new functionality gets to be `make_list_series_with_offsets` and takes a `series` and `rebased_list_buffers` for type-safety (forcing the caller to themselves call `rebase_list_offsets` and thereby understand that its correct.

IyeOnline commented on libtenzir/builtins/operators/ocsf.cpp:
```diff
@@ -1200,6 +1199,102 @@ class deriver {
+
+  auto same_list_layout(const arrow::ListArray& lhs,
+                        const arrow::ListArray& rhs) -> bool {
+    if (lhs.length() != rhs.length() or lhs.null_count() != rhs.null_count()) {
```
I believe it should be impossible for the `ListArray`'s `length()` to differ, since they are columns in the same slice.

IyeOnline commented on libtenzir/builtins/operators/ocsf.cpp:
```diff
@@ -1200,6 +1199,102 @@ class deriver {
+    if (lhs.length() != rhs.length() or lhs.null_count() != rhs.null_count()) {
+      return false;
+    }
+    const auto lhs_base = lhs.value_offset(0);
```
This should use `value_length(i)` instead of doing maths itself.

mavam committed:
Clarify list series rebuilding

Split the unsafe full-values rejoin path from the rebased-offset path for sliced list arrays. Also use Arrow list accessors when comparing layouts and record the convention in the agent references.

Assisted-by: ChatGPT (pi)
